internal/observability/generic.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package observability sets up and configures observability tools.
+package observability
+
+import (
+	"sync"
+
+	"go.opencensus.io/plugin/ocgrpc"
+	"go.opencensus.io/plugin/ochttp"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/trace"
+)
+
+// Compile-time check to verify implements interface.
+var _ Exporter = (*GenericExporter)(nil)
+
+var initExporterOnce sync.Once
+
+type traceAndViewExporter interface {
+	trace.Exporter
+	view.Exporter
+}
+
+// GenericExporter is a standard implementation of an exporter that wraps the opencensus interfaces
+// with custom configuration
+type GenericExporter struct {
+	exporter   traceAndViewExporter
+	sampleRate float64
+}
+
+func (g *GenericExporter) InitExportOnce() {
+	initExporterOnce.Do(g.initExporter)
+}
+
+func (g *GenericExporter) initExporter() {
+	if g.exporter == nil {
+		return
+	}
+	trace.ApplyConfig(trace.Config{
+		DefaultSampler: trace.ProbabilitySampler(g.sampleRate),
+	})
+	trace.RegisterExporter(g.exporter)
+	view.RegisterExporter(g.exporter)
+
+	// Record the various HTTP view to collect metrics.
+	httpViews := append(ochttp.DefaultServerViews, ochttp.DefaultClientViews...)
+	if err := view.Register(httpViews...); err != nil {
+		panic(err)
```

Comment:
Sorry I missed this in the first pass. I don't think we should panic, especially since we control when `InitExportOnce` is called. Can we change the interface to return an error and catch that error in setup?

Reply:
(Yes, this was the behavior before, but while we're refactoring... 😄)

Reply:
Thanks @sethvargo - thats not as simple as it seems because the `sync.Once`'s `.Do()` function wraps the initExporter() and does not return an error.

Reply:
Guess I'll just grab the err from the .Do and return it anyways. Should be fine. PR coming shortly.

Reply:
See https://github.com/google/exposure-notifications-server/pull/646